hls-plugin-api/src/Ide/Types.hs

```diff
@@ -389,6 +389,7 @@ data PluginCommand ideState = forall a. (FromJSON a) =>
 
 type CommandFunction ideState a
   = ideState
+  -> PluginId
```

Comment:
I'm not sure this is an improvement. The other arguments to `CommandFunction` are things that are only known when the command is run, so we need to pass them in then. But the `PluginId` is known when the plugin is created, so we can easily pass it in there. Doing it that way is pretty straightforward, as the `tactics` plugin does today.

Reply:
Plus, most of the existing ones don't use it, so we've also added a bunch of unused parameters...

Reply:
Only the command parameter is needed to be passed, I think `ideState` has the same condition as `pluginId`, we can also pass `ideState` while we only need it.

---
Here are the stat about the usage of `CommandFunction`.

8/11 use ideState, 3/11 use PluginId, and most of plugins don't use `PluginId` because they use `MaybeT` to wrap `NormalizedFilePath` conversion, I keep this pattern in the pr. I will respect your view if you still believe this change is of little significance.

plugin name | ideState | PluginId
------ | ------ | -----
completion | ✓ | ×
typelense | ×| ×
class |  ✓| ×
eval |  ✓|✓
explicit-import |× |×
hlint |  ✓| ✓
module-name |  ✓|×
refine-import | ×|×
retrie |  ✓|×
splice |  ✓|×
tactics |  ✓|✓

Reply:
If I may add, like I mentioned earlier I want to clean up all of the duplicate code revolving around MaybeT. So I would plan on removing the MaybeT wrappers. It may end up being more daunting but I expect to need to use some sort of identifier to use in logging messages. 

This would also extend to other commonly requested items as well. So it would be nice to have some generic identifier to pass around.

Reply:
TBH I also think it looks like we shouldn't pass `IdeState` like that, if it truly is globally constant. I'm a bit suspicious that it isn't truly constant though, we should check.

I think my argument is that it's unclear what should be part of `CommandFunc`. Just any convenient stuff? Then there's really no rule about what to put in there or not. Whereas with my proposal there is a simple rule: it should take exactly the stuff that's only available when you run the command, and nothing else. If you need anything else, just pass it in when you construct the plugin.

Reply:
> I'm a bit suspicious that it isn't truly constant though, we should check.

I think my changes have covered all usage of `CommanFunc`.

> I think my argument is that it's unclear what should be part of CommandFunc. Just any convenient stuff? Then there's really no rule about what to put in there or not.

I believe `CommandFunc` is an abbreviation of its arguments and its result, and the arguments can be anything you may use.
I prefer to add something that may not be used to keep the `CommandFunc` be unified in most situations.

> So I would plan on removing the MaybeT wrappers.

Agree with this.

--- 

As I said earlier, I'm going to close this and open a new pr to make getNormalizedFilePath takes Uri as pram instead of TextDocumentIdentifier only.